README.txt
dateable.chronos
================

Dateable.chronos is a set of calendar views for Plone.

Features
--------

Day, Week and Month calendar views. Also an upcoming list and past list.


Installation
------------

Installation is done via the installer of the bundling product, i.e. 
p4a.plonecalendar.


Requirements
------------

* dateable.kalends

* Plone.

* It's currently developed as a part of p4a.plonecalendar, but is in theory
  usable with other integrations with Plone.
  

CHANGES.txt
=========================
dateable.chronos Changes
=========================

1.0.1 (unreleased)
==================

- A link in the popup was incorrect.


1.0 (2013-05-21)
================

- Include zope.security zcml instead of defining the
  zope.ManageContent permission ourselves.
  [maurits]

- Plone 4.3 compatibility. [regebro]

- CSS fixes, and template cleanup. [regebro]

This version was tested with Plone 4.0.10, 4.1.6, 4.2.5 and 4.3.

0.8 (2012-08-04)
================

- The views are now normal views for Plone Folderish things. This is so we
  can get rid of p4a.subtyper dependencies p4a.plonecalendar. This means
  dateable.chronos now is dependant on Plone, but afaik it's only used in
  Plone anyway. [regebro]
  
- Got rid of version.txt and MANIFEST.in [regebro]

dateable.chronos 0.7.2.2 (2012-07-10)
=====================================

- include also doc/ with distribution (added in MANIFEST.in)
  [iElectric]

dateable.chronos 0.7.2.1 (2012-07-10)
=====================================

- include version.txt with distribution (added MANIFEST.in)
  [iElectric]

dateable.chronos 0.7.2 (2012-06-26)
===================================

- fix AttributeError in base_view.py
  [iElectric]

- Do not (silently) depent on p4a.common, but have the required timezone code bit locally
  [miohtama]

- make views subclass Products.Five.browser.BrowserView so that their templates
  can be customized through portal_view_customizations
  [swampmonkey]

- Fixed a XSS vulnerability in the get_view_day method
  [keul]

dateable.chronos 0.7.1 (2010-08-01)
===================================

* Import getMultiAdapter from zope.component and not zope.app.zapi.
  This works with Zope >= 2.10 [tom_gross]

* Added Dutch translation.
  [WouterVH]


dateable.chronos 0.7.0 (2009-11-27)
===================================

* Added Month/Week/Day view attributes on the event list. Also added
  IListCalendarEnhanced interface. If you mark a calendar with 
  IListCalendarEnhanced instead of ICalendarEnhanced, then these new list
  views will be used for Month/Week/Day views instead of the grid ones.
  [regebro]
  
* Got rid of that annoying '-C' that pops up in some URLs when you navigates.
  [regebro]
  
* Dropped official Plone 3.0 and 3.1 support. It probably still works, but
  the buildouts have for some reason stopped working.
  [regebro]

dateable.chronos 0.6.2 (2009-09-22)
===================================

* Added Czech translation.
  [naro]

* Added Japanese translation.
  [tyam]

* Added Turkish translation (in fact this was added to 0.6.1, but not noted
  in the changes.)
  [tyam]

* Corrections to the italian translation.
  [massimo]
  
* The KSS JS is now enable also for anonymous users.
  [regebro]

dateable.chronos 0.6.1 (2009-06-30)
===================================

* Added Italian translation.
  [massimo]
  
* Added Polish Translation by Magdalena Regebro and Danish translation by
  Alexander Pilz.
  [regebro]
  
* Updated translations as best I could. 
  [regebro]
  
* The localization of the time format now uses the localization support for
  date and time formats in zope.i18n.
  [regebro]

dateable.chronos 0.6 (2009-06-25)
=================================
* Added norwegian translation by Mats Tande <mats@tande.net>. 
  [regebro]
  
* Put in a larger space between the List-tabs and the other tabs as a
  distinguishing marker. This is in CSS so it can be easily overridden.
  [regebro]

* Added a class onto a day of the month view if that day has events
  [cwainwright]

* Properly localized the format of hours shown on the Day and Week views.
  [erikrose,catherine-w]
  
* Testing improvements.
  [erikrose]

dateable.chronos 0.5 (2009-05-03)
=================================

* Fixed an issue where utils.get_view_day would choke on a day that
  didn't exist in the month.
  [claytron]
  
* Several improvements of i18n including German translation.
  [xo7,paulox]  

dateable.chronos 0.4 (2008-10-04)
=================================

* fixed the getEventDisplays method for day.py to use the 
  getOccurrencesInDay method since it already does the work for us
  and it passes in the request.form so we can modify the search.
  [claytron]

* fixed the calendar_view_tabs macro so that the id gets the proper
  title from the action, thus validating the page.
  [claytron]

* made the menu items actions (day, month, week, list, past) keep
  what we had in the request
  [lucielejard]

* when we get the occurrences in a day, made the day end at 11:59pm
  instead of 12am the next day. This was causing events starting at
  12am to start displaying the day before.
  [lucielejard]

* hooked up the selection of the hour height for the day view
  [lucielejard]

* hooked up the selection of how many events we want to display
  on the month view and if we even want to have a limit
  [lucielejard]

* hooked up the selection of "start hour" and "end hour" 
  (site setup -> chronos configuration) with the day and week views. 
  [lucielejard]
 
* updated the display of events that last more than one day (on day and 
  week views). This is possible thanks to the variable column_day that
  allows you to know on which day you are displaying the event, and then
  to know where to start/stop displaying the event on this day.
  [lucielejard]
 
* introduced two variables, before_line and after_line, to know
  if we have to display the gray lines at the top and bottom of the
  day and month views. For example, if your day starts at 12am, you 
  don't need the grey line at the top, since there won't be any event
  before 12am on this day.
  [lucielejard]
 
* allowed to select "24" as the end time on the manage calendar
  form, to be able to display hours until midnight on day and
  month views.
  [lucielejard]

* Fixed the _get_view_url_for_date so that it properly checks the days in the
  month.  http://plone.org/products/dateable/issues/7
  [claytron]

dateable.chronos 0.3 (2008-07-09)

dateable/chronos/browser/week.py
# -*- coding: ISO-8859-15 -*-
# (C) Copyright 2005 Nuxeo SARL <http://nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# $Id: week.py 47783 2006-08-02 11:34:03Z lregebro $
"""
  week.py
"""

from datetime import timedelta, date
from zope.interface import implements
from zope.component import getMultiAdapter

from Products.CMFCore.utils import getToolByName

from dateable.chronos import isoweek
from base_view import BaseCalendarView
from displaytable import DayGrid
from interfaces import IPositionedView, IEventDisplay

from zope.i18nmessageid import MessageFactory
_ = MessageFactory("calendar")

# Some defaults for rendering:
HOUR_HEIGHT = 30 # Two minutes per pixel
CALENDAR_WIDTH = 738

class WeekView(BaseCalendarView):
    """Holds the rendering information for week views"""
    
    implements(IPositionedView)

    def prev_long_link(self):
        return self.prev_month_link()

    def prev_short_link(self):
        return self.prev_week_link()

    def next_short_link(self):
        return self.next_week_link()

    def next_long_link(self):
        return self.next_month_link()

    def getDateForWeekday(self, weekday):
        days_before = self.daysUntil(self.firstweekday, self.default_day.weekday())
        week_begins = self.default_day - timedelta(days_before)
        
        offset = self.daysUntil(self.firstweekday, weekday)
        return week_begins + timedelta(offset)
    
    def calcInfo(self):
        self.calendar = self.context
        self.year = int(self.year())
        self.first_day = self.column_day = self.getDateForWeekday(self.firstweekday)
        self.month = self.first_day.month
        # self.ends is the day after the last day
        self.ends = self.first_day + timedelta(8)
        self.days = 7 

        p_cal = getToolByName(self.context, 'portal_calendar')
        self.from_hour = p_cal.hour_start
        self.to_hour = p_cal.hour_end
        self.hour_count = self.to_hour - self.from_hour

        # We might want to make this more flexible for different layouts
        # The 650 pixel width is good for the default CPS design on my
        # screen. Not very generic.
        self.day_width = CALENDAR_WIDTH / self.days
        # Make sure they match even if calendar_width isn't a multiple of 7
        self.width = self.day_width * self.days

        self.hour_height = HOUR_HEIGHT
        self.height = self.hour_height * self.hour_count
        # Add on the "before" and "after" lines.
        self.before_line = False
        self.after_line = False
        if self.from_hour > 0:
            self.height = self.height  + self.hour_height 
            self.before_line = True
        if self.to_hour < 23:
            self.height = self.height  + self.hour_height 
            self.after_line = True

    def getDays(self):
        return [self.getDateForWeekday(d) 
                for d in self.getWeekdayNumbers()]
            
    def getEventDisplays(self):
        all_displays = []
        for d in self.getWeekdayNumbers():
            day = self.getDateForWeekday(d)
            self.column_day = day
            occurrences = self.getOccurrencesInDay(day)
            displays = [getMultiAdapter([occurrence, self], IEventDisplay) for
                        occurrence in occurrences]
            
            # Handle overlapping displays:
            daygrid = DayGrid(0, self.day_width * (self.daysUntil(self.firstweekday, d)), 
                                 self.height,
                                 self.day_width)
            daygrid.extend(displays)
            daygrid.flatten()
            all_displays.extend(daygrid)
            
        return all_displays

    def getTodayInfo(self):
        today = date.today()
        if (today >= self.first_day.date() and 
            today < (self.first_day + timedelta(7)).date()):
            istoday = True
        else:
            istoday = False
        return {'year': today.year, 
                'month': today.month, 
                'day': today.day,
                'istoday': istoday}
 
    def getPreviousViewUrl(self):
        return self.getDateUrl(self.default_day - timedelta(days=7), 'week')

    def getNextViewUrl(self):
        return self.getDateUrl(self.default_day + timedelta(days=7), 'week')
    
    def getTodayUrl(self):
        return self.context.absolute_url() + '/week.html?date='
